### ISSUE TITLE
Labels of factors lost

### ISSUE THREAD

**Issue by stefvanbuuren** [stefvanbuuren/mice#25]
Adam Nieuzytek alerted me of the following problem:

```r
## labeling bug 27feb2017

library("mice")
imp <- mice(nhanes2)
cmp <- complete(imp)
head(cmp) ## does have lavels, just data

# here we lose the the category labels
# because this is a contrast, not a category anymore
fit <- lm(bmi~age+hyp+chl, data = cmp)
fit

# however, the following is fine
fit2 <- lm( bmi~age+hyp+chl, data = nhanes2)
fit2

# there is a difference between
str(nhanes2)
str(cmp)
```
Presumably, somewhere in `complete()` there is a problem that creates this difference.

**Comment by stephematician** [comment]
The `data` member of the mids object returned by `mice()` seems to be the origin of the changes to the factors, rather than the `complete()` function. This is because the `p$data` object is used to initialise the `data` member of the returned mids object.

It isn't clear to me why (subsetted) `p$data` is used to construct the returned object, because

  1. the `pad` member (of the returned mids object) is initialised with `p`, so `p$data` is certainly available to any functions that will need it, and
  2. I think `p$data[,seq_len(nvar)]`, aside from the contrast attribute attached to any column which is a factor, should be exactly the same as `data`.

If you debug(mice) before running the above, the states of `data` and `p$data[,str_len(nvar)]` before the final statement in `mice()` are:

    Browse[2]> str(data)
    'data.frame':   25 obs. of  4 variables:
     $ age: Factor w/ 3 levels "20-39","40-59",..: 1 2 1 3 1 3 1 1 2 2 ...
     $ bmi: num  NA 22.7 NA NA 20.4 NA 22.5 30.1 22 NA ...
     $ hyp: Factor w/ 2 levels "no","yes": NA 1 1 NA 1 NA 1 1 1 NA ...
     $ chl: num  NA 187 187 NA 113 184 118 187 238 NA ...
    Browse[2]> str(p$data[,seq_len(nvar)])
    'data.frame':   25 obs. of  4 variables:
     $ age: Factor w/ 3 levels "20-39","40-59",..: 1 2 1 3 1 3 1 1 2 2 ...
      ..- attr(*, "contrasts")= num [1:3, 1:2] 0 1 0 0 0 1
      .. ..- attr(*, "dimnames")=List of 2
      .. .. ..$ : chr  "20-39" "40-59" "60-99"
      .. .. ..$ : chr  "2" "3"
     $ bmi: num  NA 22.7 NA NA 20.4 NA 22.5 30.1 22 NA ...
     $ hyp: Factor w/ 2 levels "no","yes": NA 1 1 NA 1 NA 1 1 1 NA ...
     $ chl: num  NA 187 187 NA 113 184 118 187 238 NA ...

I also can't think of a situation in which this wouldn't be true, so surely `data=data` in the initialisation of the mids object would make sense?

**Comment by stephematician** [comment]
One caveat to the comment about `pad` being initialised to `p` ... the `pad` object is deleted if `diagnostics==F` - in which case the contrast information is not stored anywhere in the returned mids object (except in `data` given the code as it is now)

**Comment by stefvanbuuren** [comment]
Fixed in 2.46.9
